plugins/woocommerce/includes/admin/importers/class-wc-product-csv-importer-controller.php

```diff
@@ -113,37 +115,18 @@ public static function is_file_valid_csv( $file, $check_path = true ) {
 	 * @throws \Exception When file validation fails.
 	 */
 	protected static function check_file_path( string $path ): void {
-		$is_valid_file = false;
+		$wp_filesystem = FilesystemUtil::get_wp_filesystem();
```

Comment:
Interesting side-effect of this. First, suppose that the following are true:

- `FS_METHOD` is not defined anywhere. That to say, the site operator hasn't added a definition to their `wp-config.php` or anything like that, and we're not in a hosting environment (like JN) where it's defined for us.
- `wp-content/` is not writable, even though specific sub-directories like `wp-content/uploads/` *are* writable ... *or,* alternatively, it is owned by a different user (than the current web server user).

In those conditions:

- We try to initialize the WP Filesystem.
- WP will try to figure out the best method to use, by writing a [temporary file to the `wp-content/` dir](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-admin/includes/file.php#L2260-L2261). This will fail given either of the permissions scenarios noted above.
- There are various possibilities from here, but perhaps the most likely outcome is it will [try to fall back on the FTP method (`ftpext`)](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-admin/includes/file.php#L2297-L2299).
- Back in our `FilesystemUtil` we try to get [filesystem (FTP) credentials](https://github.com/woocommerce/woocommerce/blob/9.3.2/plugins/woocommerce/src/Internal/Utilities/FilesystemUtil.php#L86) which means WP will (attempt to) *display a form to the user to request for their FTP/SSH details* and, unfortunately, that flow is not going to work here/we're not interfacing with it correctly.
- Result is an [exception is thrown](https://github.com/woocommerce/woocommerce/blob/9.3.2/plugins/woocommerce/src/Internal/Utilities/FilesystemUtil.php#L27), which is then presented to the user:

![wpfs-could-not-init](https://github.com/user-attachments/assets/4a5ecdbd-511d-4008-871b-cce92536833d)

This does not happen with the current stable release or with trunk, it's only because we're introducing a call to `FilesystemUtil::get_wp_filesystem()` into this flow and, again, would impact any sites where the `FS_METHOD` is not explicitly configured and where file permissions are as described (I'm a little unsure how common that would be, but it also doesn't sound especially atypical to me).

Reply:
Hummm maybe we could use the WP's `filesystem_method` filter to assume "direct" as the fallback method when `FS_METHOD` is not defined?

Reply:
Possibly ... though it might require some other checks, too. 

For instance, it seems `FS_METHOD` could be undefined, but `ftpext` may be the correct and valid fallback if teh [`ftp_credentials` option is set](https://github.com/WordPress/wordpress-develop/blob/6.6.2/src/wp-admin/includes/file.php#L2414-L2420) or if a [number of FTP constants are set](https://github.com/WordPress/wordpress-develop/blob/6.6.2/src/wp-admin/includes/file.php#L2436-L2442).

Then again, we were effectively assuming `FS_METHOD` := `'direct'` before this change (as we were not using WP Filesystem), so I think it may still be a good way forward 🤔

Reply:
> Back in our FilesystemUtil we try to get [filesystem (FTP) credentials](https://github.com/woocommerce/woocommerce/blob/9.3.2/plugins/woocommerce/src/Internal/Utilities/FilesystemUtil.php#L86) which means WP will (attempt to) display a form to the user to request for their FTP/SSH details and, unfortunately, that flow is not going to work here/we're not interfacing with it correctly.

One clarification here, we avoid the FTP credentials form and related error messages by wrapping the `request_filesystem_credentials` call in an output buffer (which is something WP Core also does in some cases). This means we will only retrieve the credentials if they are set in the option or the constants, and we never ask for them from the user directly.

Reply:
Maybe in our `initialize_wp_filesystem` method we should allow for specifying the directory being tested? The second argument of `get_filesystem_method` is `$context`, which is then used to to the file write test. So if, like the scenario above:

> wp-content/ is not writable, even though specific sub-directories like wp-content/uploads/ are writable ... or, alternatively, it is owned by a different user (than the current web server user)

...we could specify the uploads directory as the context, perhaps we could avoid some cases where the "could not be initialized" exception is thrown?

Reply:
That sounds like it could work nicely.

Reply:
So it would be like `function initialize_wp_filesystem($dir_to_check='')` and then `initialize_wp_filesystem(wp_get_upload_dir()['basedir'])`?

Reply:
🤔 That's kind of what I was thinking. But the value for `$dir_to_check` would either have to be passed through from the `get_wp_filesystem` method, or we'd have to make `initialize_wp_filesystem` public. I guess this could get complicated, because there's a scenario where you might want to use the WP_Filesystem class in two different directories that might have different permissions, so you'd need to be able to re-initialize when switching to the other directory...

Reply:
> Then again, we were effectively assuming FS_METHOD := 'direct' before this change (as we were not using WP Filesystem), so I think it may still be a good way forward

What if we just do this and take a reactive approach moving forward? I mean, if we get some report of the lack of the `FS_METHOD` constant being a problem, then (and not before) we think how to solve that.

Reply:
Just to be extra clear 🙌🏼  ... you mean *using WP's `filesystem_method` filter to assume "direct" as the fallback method when FS_METHOD is not defined*?

Yep, I think that may be pretty reasonable, except we should probably expand it to cover those other scenarios I mentioned, so, more like: *using WP's `filesystem_method` filter to assume "direct" as the fallback method when FS_METHOD is not defined **and** [ftp_credentials option](https://github.com/WordPress/wordpress-develop/blob/6.6.2/src/wp-admin/includes/file.php#L2414-L2420) is not set **and** [the FTP constants](https://github.com/WordPress/wordpress-develop/blob/6.6.2/src/wp-admin/includes/file.php#L2436-L2442) (or a single representative constant) are not set.*

Reply:
So to be extra-extra clear... 🙂

```
$method = get_filesystem_method();
if(!$method) {
  if(no FTP credentials && no FTP constants) {
    assume "direct"
  }
  else {
    ...assume FTP? Then 'ftpext'? or 'ftpsockets'?
  }
}
```

Right?

Note that we don't need to check for `FS_METHOD` explicitly, since `get_filesystem_method()` will already do that.

Reply:
*Yes* (at least, that's what I'm picturing).